examples/ctc.dx

```diff
@@ -30,115 +33,115 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
-      (blank:  vocab)
-      (logits: time=>vocab=>Float)
-      (labels: position=>vocab)
-      : Float =
-  -- Computes log p(labels | logits), marginalizing over possible alignments.
-  -- Todo: remove unnecessary implicit type annotations once
-  -- Dex starts putting implicit types in scope.
+def prepend_and_interleave {m v} (blank:v) (seq: m=>v) : ((Unit|(m & IsBlank))=>v) =
+  -- Turns "text" into " t e x t ".
+  prepend blank (interleave blank seq)
 
-  ilabels = prepend_and_interleave blank labels
-
-  normalized_logits = for t. logsoftmax logits.t
-
-  -- Initialization rules
-  logprob_start_with_blank = normalized_logits.(0@_).blank
-  logprob_of_first_label   = normalized_logits.(0@_).(labels.(0@_))
-  log_prob_seq_t0 = for pos.
-    -- TODO: allow pattern-matching on integer literals
-    case ordinal pos == 0 of
-      True -> logprob_start_with_blank
-      False -> case ordinal pos == 1 of
-        True  -> logprob_of_first_label
-        False -> log 0.000001
-
-  same_as_last = \ilabels s.
-    o = n_to_i $ ordinal s
-    select (o >= 2) (ilabels.s == ilabels.(clipidx _ (o - 2))) False
-
-  safe_idx = \prev s.
-    select (s >= 0) prev.(clipidx _ s) (log 0.0)
-
-  -- Todo: As suggested by Adam Paske, we could get rid of these
-  -- logaddexp calls with a newtype that overloads + and *
-  update = \t prev.
-    case ordinal t == 0 of
-      True -> log_prob_seq_t0
-      False -> for s.
-        cond = ilabels.s == blank || same_as_last ilabels s
-        labar = logaddexp prev.s (safe_idx prev (n_to_i (ordinal s) - 1))
-        other = logaddexp labar  (safe_idx prev (n_to_i (ordinal s) - 2))
-        ans = select cond labar other
-        ans + normalized_logits.t.(ilabels.s)
+def normalize {n v} [Fractional v, Add v] (xs:n=>v) : n=>v =
+  for i. divide xs.i (sum xs)
 
-  log_prob_seq_final = fold log_prob_seq_t0 update
 
-  -- Todo: nicer way to get last two elements of log_prob_seq_final.
-  seq_length = 1 + size (position & (Fin 2))
-  endlabel = log_prob_seq_final.((unsafe_nat_diff seq_length 2)@_)
-  endspace = log_prob_seq_final.((unsafe_nat_diff seq_length 1)@_)
-  logaddexp endlabel endspace
+'### Main CTC algorithm
+Computes p(labels | logits), marginalizing over possible alignments and
+and marginalizes over insertion of blanks between characters.
+Todo: add a `[Subset position time]` constraint.
 
+def ctc_inner {vocab time position} [NonEmpty vocab, NonEmpty position, NonEmpty time, Eq time, Eq vocab]
+    (blank: vocab)
+    (logits: time=>vocab=>LogSpace Float)
+    (labels: position=>vocab)
+      : LogSpace Float =
+  ilabels = prepend_and_interleave blank labels
+  normalized_logits = for t. normalize logits.t
+
+  -- Initialize dynamic programming table.
+  logprob_start_with_blank = normalized_logits.first_ix.blank
+  logprob_of_first_label   = normalized_logits.first_ix.(labels.first_ix)
+  log_prob_seq_t0 = yield_state (for pos. Exp (-10000.0)) \lpRef.
```

Comment:
You can't write -infinity instead of -10000 here?  Is there a NaN bug somewhere in the prelude already?

Reply:
Yeah that bugs me too.  But in the main loop we have to multiply two probabilities, which I think is where the `-infinity` can turn into a `NaN`.

Reply:
Shouldn't.  This example works:

```
import stats

Exp (-infinity) * Exp (-infinity)
> (Exp -Infinity)
```

If there's something that doesn't we should probably track it down and fix it.

Reply:
Thanks to @axch 's help I did track this down and made our `logsumexp` functions handle infinities correctly.